lib/puppet/pal/pal_impl.rb

```diff
@@ -486,6 +507,38 @@ def self.main(
   end
   private_class_method :main
 
+  # Warn and remove variables that will be shadowed by facts of the same
+  # name, which are set in scope earlier.
+  def self.merge_vars(target_vars, vars, facts)
+    # First, shadow plan and target variables by facts of the same name
+    vars        = shadow_vars(facts || {}, vars, 'fact', 'plan variable')
+    target_vars = shadow_vars(facts || {}, target_vars, 'fact', 'target variable')
+    target_vars = shadow_vars(vars, target_vars, 'plan variable', 'target variable')
+
+    # Then, shadow target variables by plan variables of the same name
```

Comment:
I think this comment belongs before #516 - by this point, target vars and plan vars shouldn't share any keys since we filtered out target variables that collide with plan variables.

Reply:
Ah, good catch 😅